src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -64,10 +64,51 @@ void parallel_reduce (const TeamMember& team,
 
    result = local_tmp;
   } else {
-    Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team, begin, end), lambda, result);
+    Kokkos::parallel_reduce(loop_boundaries, lambda, result);
   }
 }
 
+template <bool Serialize, int TotalSize, int VectorSize, typename TeamMember, typename InputProvider, typename ValueType>
```

Comment:
I think you could deduce VectorSize from the return type of `input(i)`. Inside the body, you could just do

```
using PackType = decltype(input(0));
constexpr int VectorSize = PackType::n; // only works if PackType is a pack
```
or, if you want to be able to handle also views of not pack types,
```
using ValueType = decltype(input(0));
constexpr int VectorSize = sizeof(PackType) / sizeof(ScalarTraits<PackType>::scalar_type);
```
Note: it might be a bit confusing if ValueType is not a Pack (talking of VectorSize for a double seems confusing). So if you DON'T want this function to be used unless the input provider provides Pack's (e.g., a View<Pack<Real,N>*>), you could make the signature like this:
```
template< [...] >
static KOKKOS_INLINE_FUNCTION
auto view_reduction( [...] ) -> OnlyPackReturn<decltype(input(0)),void>
```
so that the function is not considered by the compiler unless input(k) returns a Pack.

Reply:
I had issues with implementing all these suggestions. For the time being, I have just added `vector_size` as another input parameter.

From the compilation errors, `using PackType = decltype(input(0));` seems to be recognizing the Pack type (`Pack<double, 3>` for example), but then the compiler was saying 
`n` is not a member of `Pack<double, 3>&` 
when calling `PackType::n`. 
Similarly, any call `npack<PackType>(TotalSize);` or `ekat::npack<PackType>(TotalSize);` said it could not deduce the template parameters. Not really sure why the ekat_pack objects aren't being recognized, especially since `reduce_sum()`, defined in the same file, is recognized. Other than it does not have any explicitly defined template params, which must be the difference.

Any suggestions? Or is `vector_size` ok as an input parameter alongside `total_size`.? Let me know if I need to give a better explanation..

Reply:
Ah, it looks like you need to remove the reference:

```
using PackType = typename std::remove_reference<decltype(input(0))>::type;
```

Reply:
Yep, just found that